%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%

%%%CONTEMPORARY%%%
\documentclass[unnumsec,webpdf,contemporary,large]{oup-authoring-template}%
%\documentclass[unnumsec,webpdf,contemporary,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,contemporary,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,contemporary,small]{oup-authoring-template}

%%%MODERN%%%
%\documentclass[unnumsec,webpdf,modern,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,modern,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,small]{oup-authoring-template}

%%%TRADITIONAL%%%
%\documentclass[unnumsec,webpdf,traditional,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,traditional,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,namedate,webpdf,traditional,medium]{oup-authoring-template}
%\documentclass[namedate,webpdf,traditional,small]{oup-authoring-template}

%\onecolumn % for one column layouts

%\usepackage{showframe}

\graphicspath{{Fig/}}

% line numbers
%\usepackage[mathlines, switch]{lineno}
%\usepackage[right]{lineno}
\usepackage{lipsum}

\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}%
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.
\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%
\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}

\begin{document}

\journaltitle{Journal Title Here}
\DOI{DOI HERE}
\copyrightyear{2022}
\pubyear{2019}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}

\firstpage{1}

%\subtitle{Subject Section}

\title[Short Article Title]{Article Title}

\author[1,$\ast$]{First Author}
\author[2]{Second Author}
\author[3]{Third Author}
\author[3]{Fourth Author}
\author[4]{Fifth Author\ORCID{0000-0000-0000-0000}}

\authormark{Author Name et al.}

\address[1]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[2]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[4]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}

\corresp[$\ast$]{Corresponding author. \href{email:email-id.com}{email-id.com}}

\received{Date}{0}{Year}
\revised{Date}{0}{Year}
\accepted{Date}{0}{Year}

%\editor{Associate Editor: Name}

%\abstract{
%\textbf{Motivation:} .\\
%\textbf{Results:} .\\
%\textbf{Availability:} .\\
%\textbf{Contact:} \href{name@email.com}{name@email.com}\\
%\textbf{Supplementary information:} Supplementary data are available at \textit{Journal Name}
%online.}

\abstract{Abstracts must be able to stand alone and so cannot contain citations to
the paper's references, equations, etc. An abstract must consist of a single
paragraph and be concise. Because of online formatting, abstracts must appear
as plain as possible.}
\keywords{keyword1, Keyword2, Keyword3, Keyword4}

% \boxedtext{
% \begin{itemize}
% \item Key boxed text here.
% \item Key boxed text here.
% \item Key boxed text here.
% \end{itemize}}

\maketitle

%Characteristics of scalability and their impact on performance.
\section{Introduction}\label{sec1}
Phylogenetic network is an acyclic directed graph that generalizes the bifurcating phylogenetic tree by allowing nodes to have indegree of two, thereby creates a reticulation structure \citep{huson2010,kong2022a}. Phylogenetic networks depict complex biological scenarios that the trees cannot, such as hybrid speciation, introgression, allopolyploid speciation, and so on \citep{huson2006}[add the new review paper here]. A handful of computational methods that estimate networks from genomic data has been proposed, however, the wide use of networks in practice is hindered by their lack of scalability which refers to the ability of a system to process a growing amount of work in a decreasing or stable amount of time \citep{bondi2000}. More precisely, phylogenetic network estimation belongs to the class of NP-Hard problems (non-deterministic polynomial-time). Some attempts to ameliorate this issue has been made but the computational requirement is still much higher for the dataset size typically applied to the tree estimation (i.e., tens of taxa).

A common strategy to enhance the efficiency of the network inference is to summarize input sequence data into a set of gene trees in prior to the analysis as implemented in many functions in \textsc{PhyloNet} \citep{than2008,wen2018a} or \textsc{SNaQ} \citep{solis-lemus2016} available in Julia package \textsc{PhyloNetworks} \citep{solis-lemus2017}. Computational cost is further ameliorated by using composite likelihood (or pseudolikelihood) that involves decomposition of the network into a set of smaller problems (e.g., triplets or quartets), excecute likelihood computation on each of them, and combine them together to approximate the likelihood of the full network. This approach has been useful in both tree (e.g., \textsc{MP-EST} \citep{liu2010}) and network (e.g., \textsc{SNaQ}, \textsc{PhyNEST} \citep{kong2022c}, \textsc{PhyloNet} \citep{yu2015,zhu2018a}) inference and shown to be much faster than the full likelihood or the Bayesian methods, without compromising accuracy \citep{hejase2016}.

Nevertheless, network inference is still a computationally demanding procedure. It is not uncommon to conduct the analysis with a handful of taxa, which narrows the scope of biological investigation. In this study, we present \textsc{SNaQ2}, a new version of \textsc{SNaQ} with improved computational efficiency via parallelization of the composite likelihood computation and making probabilistic decisions during the network searching heuristics. In the following, we first briefly introduce the essence of the original \textsc{SNaQ} followed by the key improvemetns made in \textsc{SNaQ2}. Then, we present the result of our benchmarks that compares the performance of \textsc{SNaQ} and \text{SNaQ2} and we apply \textsc{SNaQ2} on empirical datasets. Our results clearly demonstrate improved efficiency in \textsc{SNaQ2}.

\section{Methods}\label{sec2}
\subsection{Original SNaQ}\label{subsec1}
As mentioned, \textsc{SNaQ} estimates phylogenetic networks from multi-locus data using composite likelihood. While detailed description of the method is available in \cite{solis-lemus2016}, we make a brief description here for readers to understand the improvemetns in \textsc{SNaQ2} in the following subsection. First, \textsc{SNaQ} extracts unrooted quarnet (i.e., networks with four tips) from the full network, and computes expected concordance factor (CF) for each quarnet that represents the proportion of genes whose true relationship is the quartet under the coalescent model. Note there are $4 \choose 2$=3 possible ways to cluster four taxa into two groups of two (i.e., separated by a split). In other words, for a taxon set $X\in\{a,b,c,d\}$ there can be three unrooted quarnets, $q_1$ that contains $a$ and $b$ on one side and $c$ and $d$ on the other, denoted by $q_1=ab|cd$, $q_2=ac|bd$, and $q_3=ad|bc$.

Given a set of estimated gene trees $G=\{G_1,G_2,\dots,G_g\}$ for $g$ loci, the number of gene trees that match with the each of the three quarnets is denoted $X=(X_{q_1},X_{q_2},X_{q_3})$. Considering $X$ follows a multinomial distribution with probabilities $(CF_{q_1},CF_{q_2},CF_{q_3})$ (i.e., the expected CF for each quarnet) assuming unlinked loci. For a level-1 network with $n\ge4$ taxa, the composite likelihood is computed using:

\begin{equation}
    L=\prod_{s \in S}(CF_{q_1})^{X_{q_1}}(CF_{q_2})^{X_{q_2}}(CF_{q_3})^{X_{q_3}}
\label{eqn1}
\end{equation}

\noindent where $S$ is the collection of all quarnets exctracted from the network.

To find the network topology that has the best fit of the observed data, the network space is searched using heuristics.  Five `moves' are implemented to traverse the networks and jump between different dimentions, which are (1) nearest-neighbor interchange (NNI), (2) addition (or deletion) of a reticulation, (3) change direction of the reticulation edge, and move (4) the target or (5) the origin of an existing hybridization edge. At each iteration, one of the randomly selected `move' makes a slight modification to the original topology $N_0$ and propose the new topology $N_1$. \textsc{SNaQ} uses hill climbing algorithm to traverse the network space and to find an optimum. So if the composite likelihood of $N_1$ is greater than $N_0$, $N_0$ is discarded and $N_1$ becomes $N_0$; but the move does not result in $N_1$ with improved composite likelihood, $N_1$ is discarded and another move is applied to $N_0$. This process continues until an optima is reached. 

Since hill climbing only guantees to find the local optimum rather than global optimum, \textsc{SNaQ} executes multiple independent runs, which is typical practice in phylogenetic heuristics. Furthermore, starting topology is modified at some probablility at the onset of the search for exhaustive search. 

\subsection{Improvements in SNaQ2}\label{subsec2}
\subsubsection{Parallelization of the pseudolikelihood calculation}\label{subsubsec1}
Original \textsc{SNaQ} makes use of parallelization mechanism by allowing each run on different processors (or cores) using Julia package \textsc{Distributed}. SNaQ2 further improves the computational speed by multithreading the pseudolikelihood calculation. In particular, extraction of quartet topologies from a network, calculation of expected CFs of the extracted quartet, and computation of quarnet likelihood are parallelized. This setting allows to allocate all runs independently on seaprate high-performance computing nodess, with each node fully utilized to parallelize the pseudolikelihood calcuation for the run it is responsible for.

\subsubsection{Sampling subset of quartets for pseudolikelihood}\label{subsubsec3}
In the orginal SNaQ, all extracted quartets were used to compute pseudolikelihood of a network. While this computation is generally efficient, it may lead to the bottleneck as the number of taxa increases since there are $n choose 4$ quartets in a network. Several studies that also utilizes quartet has shown that subsampling some quartets lead to accurate estimation of a phylogeny (e.g., SVD Quartets). In SNaQ2, we added a new argument \texttt{propQuartets} in the network inference function \texttt{snaq!}. The argument \texttt{propQuartets} specifies the proportion of randomly sampled quartets for the pseudolikelihood calculation (i.e., nonnegative float equal to or smaller than 1). 

\subsubsection{Proposals using quartet weighting}\label{subsubsec2}
 All moves, except the move that changes direction of the selected reticulation edge, involves random selection of a tree edge in $N_0$ that will be modified. For example, to move the origin on an existing reticulation edge, a reticulation edge whose head is at a randomly selected reticulation node $u$ is selected at prorbability of $\gamma$, followed by a $random$ selection of a tree edge that will have a new reticulation node in the middle. This stochasticity can result in increased time requires to find the global optimum during the searching process.

We make an improvement in heuristics by selecting the edge via weighted random sampling where the weight is $\Delta CF=\sum^3_{i=1}\left|X_{q_i}-CF_{q_i}\right|$, calculated for every quartet extracted from a network. It can be done for a valid edges to choose from, although not implemented in the method. The rational here is... The new variable that is added to snaq! is probQR (varied from 0.0=full random to 1.0=full weighted).






\subsection{Evaluation using simulated and empirical data}\label{subsec3}
\subsubsection{Simulation}
We evaluate the performance of SNaQ2 using simulation. A set of species networks wher each network has $n=\{10, 20, 30\}$ tips with 1 or 3 reticulations when $n=10$ and 1, 3, or 5 otherwise was generated. For each $n$, we first generated a species tree under a Yule process using R package phytools, then we sequentially added reticulation onto the topology at arbitrary position. We checked each network is level-1 considering that snaq! can only infer networks in this class, both manually and using R package SiPhyNetworks. For each species network, we set each branch length $=\{0.5,1.0,2.0\}$ colescent unit to represent high, medium, and low amount of incomplete lineage sorting.

Using Julia package PhyloCoalSimulations, we generate a set of $g\in\{300, 1000, 3000\}$ gene trees for each species network. For each gene tree, we generated multiple sequence alignment that is $10^3$ bp long, setting the scale branch parameter=0.03 and base frequency of nucleotides as A=0.3, C=0.2, G=0.2, and T=0.3 under the HKY model. The generated sequence alignment was used to estimate a gene tree using IQ-TREE 1.6.12 with the best substituion model being identified withtin the sotftware with default parameters. Gene tree estimation error was measured using python package FastMulRFS.

The set of estimated gene trees were subsequently used as an input file for network estimation using SNaQ1 and SNaQ2. For SNaQ1, a randomly selected estimated gene tree was used as the starting topology, a table of CFs computed from the set of estimated gene trees were used as input, and the true number of reticulations were provided. For SNaQ2, all parameter setting was identical to SNaQ1 with additional parameters probQuartets$\in\{1.0, 0.9, 0.7\}$ and proqQR$\in\{0,0.5,1.0\}$. We recorded runtime for each network analysis. We evaluated the accuracy of the estimated network using hardwired cluster dissimilarity metric in Julia package PhyloNetworks. More specifically, we compared between the estimated network with the true network as well as the major trees of the estimated network and the true network. 

All computations were done using condor at University of Wisconsin-Madison. We ran the analyses in different computing power setting the number of processors $\in\{4,8,16\}$ to compare the efficiency in various conditions. One hundred replicates were made.


%Add species networks used in supplementary materials 
%http://blog.phytools.org/2015/10/simulating-species-tree-from-genus-tree.html
%Add sophisticated simulation pipeline in supp.

\subsubsection{Empirical data}

Empirical data: 1. the fish data in snaq1; 2. find something else

\section{Results and discussion}\label{sec3}
\subsection{Simulation}\label{subsec4}
GTEE
\subsection{Empirical data}\label{subsec5}

\section{Conclusion}\label{sec4}

\section{Competing interests}
No competing interest is declared.

\section{Author contributions statement}
NK and SK designed and consucted analysis, prepared the manuscript. TC implemented improvements into SNaQ. 

\section{Acknowledgments}
WID Server? 

%\bibliographystyle{plain}
%\bibliography{reference}
%USE THE BELOW OPTIONS IN CASE YOU NEED AUTHOR YEAR FORMAT.
\bibliographystyle{abbrvnat}
\bibliography{reference}

\end{document}
