%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%

%%%CONTEMPORARY%%%
\documentclass[unnumsec,webpdf,contemporary,large]{oup-authoring-template}%
%\documentclass[unnumsec,webpdf,contemporary,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,contemporary,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,contemporary,small]{oup-authoring-template}

%%%MODERN%%%
%\documentclass[unnumsec,webpdf,modern,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,modern,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,small]{oup-authoring-template}

%%%TRADITIONAL%%%
%\documentclass[unnumsec,webpdf,traditional,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,traditional,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,namedate,webpdf,traditional,medium]{oup-authoring-template}
%\documentclass[namedate,webpdf,traditional,small]{oup-authoring-template}

%\onecolumn % for one column layouts

%\usepackage{showframe}

\graphicspath{{Fig/}}

% line numbers
%\usepackage[mathlines, switch]{lineno}
%\usepackage[right]{lineno}
\usepackage{lipsum}

\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}%
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.
\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%
\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}

\begin{document}

\journaltitle{Journal Title Here}
\DOI{DOI HERE}
\copyrightyear{2022}
\pubyear{2019}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}

\firstpage{1}

%\subtitle{Subject Section}

\title[SNaQ2]{SNaQ2: Improved Scalability for Phylogenetic Network Inference}

\author[1]{Sungsik Kong}
\author[1,2]{Nathan Kolbow}
\author[3]{Tyler Chafin}
\author[1,4,$\ast$]{Claudia Solís-Lemus}

\authormark{Author Name et al.}

\address[1]{\orgdiv{Wisconsin Institute for Discovery}, \orgname{University of Wisconsin–Madison}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[2]{\orgdiv{Department of Biostatistics}, \orgname{University of Wisconsin–Madison}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[4]{\orgdiv{Department of Plant Pathology}, \orgname{University of Wisconsin–Madison}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}

\corresp[$\ast$]{Corresponding author. \href{email:email-id.com}{email-id.com}}

\received{Date}{0}{Year}
\revised{Date}{0}{Year}
\accepted{Date}{0}{Year}

%\editor{Associate Editor: Name}

%\abstract{
%\textbf{Motivation:} .\\
%\textbf{Results:} .\\
%\textbf{Availability:} .\\
%\textbf{Contact:} \href{name@email.com}{name@email.com}\\
%\textbf{Supplementary information:} Supplementary data are available at \textit{Journal Name}
%online.}

\abstract{A phylogenetic network is an acyclic directed graph that generalizes the bifurcating phylogenetic tree by allowing nodes to have an indegree of two, thereby creating a reticulation structure. Phylogenetic networks represent complex biological scenarios that phylogenetic trees cannot, such as hybrid speciation, introgression, allopolyploid speciation, and more. However, network inference is computationally demanding and often lacks scalability. It is not uncommon to conduct analysis with only a handful of taxa, which narrows the scope of biological investigation. In this study, we present SNaQ2, a new version of SNaQ—a popular summary-based network inference method that uses concordance factors. In SNaQ2, computational efficiency is enhanced through (1) weighted random selection of quartets, (2) parallelization of the quartet likelihood calculation during composite likelihood computation, and (3) probabilistic decision-making during network search heuristics. In this talk, I will first briefly introduce the essence of the original SNaQ, then describe the key improvements made in SNaQ2 in more detail. I will also present the results of our benchmarks that compare the performance of the two versions of SNaQ, along with the application of the new version to empirical datasets.
}
%Abstracts must be able to stand alone and so cannot contain citations to the paper's references, equations, etc. An abstract must consist of a single paragraph and be concise. Because of online formatting, abstracts must appearas plain as possible.}
\keywords{Composite likelihood, Parallelization, Phylogenetic Networks, Scalability}

% \boxedtext{
% \begin{itemize}
% \item Key boxed text here.
% \item Key boxed text here.
% \item Key boxed text here.
% \end{itemize}}

\maketitle

%Characteristics of scalability and their impact on performance.
\section{Introduction}\label{sec1}
Phylogenetic network is an acyclic directed graph that generalizes the bifurcating phylogenetic tree by allowing some nodes to have indegree of two and create a reticulation structure \citep{huson2010,kong2022a}. Phylogenetic networks depict complex biological scenarios that the trees cannot, such as hybrid speciation, introgression, allopolyploid speciation, and so on \citep{huson2006}[add the new review paper here]. A handful of computational methods that estimate networks from genomic data has been proposed, however, their wide use in practice is hindered by their lack of scalability (i.e., the ability of a system to process a growing amount of work in a decreasing or stable amount of time \citep{bondi2000}). More precisely, phylogenetic network estimation is am NP-Hard problem (non-deterministic polynomial-time). Some attempts to ameliorate this issue has been made but the computational requirement is still exessively high even for the dataset size typically applied to the tree estimation (i.e., tens of taxa).

A common strategy to enhance the efficiency of the network inference is to summarize input sequence data into a set of gene trees in prior to the analysis as implemented in many functions in \textsc{PhyloNet} \citep{than2008,wen2018a} or \textsc{SNaQ} (Species Network applying Quartets) \citep{solis-lemus2016} available in \textsc{Julia} package \textsc{PhyloNetworks} \citep{solis-lemus2017}. Computational cost is further ameliorated by using composite likelihood (or pseudolikelihood) that involves decomposition of the network into a set of smaller problems (e.g., triplets or quartets), excecute likelihood calculation on each of them, and combine them together to approximate the likelihood of the full network. This approach has been useful in both tree (e.g., \textsc{MP-EST} \citep{liu2010}) and network (e.g., \textsc{SNaQ}, \textsc{PhyNEST} \citep{kong2022c}, \textsc{PhyloNet} \citep{yu2015,zhu2018a}) inference and shown to be much faster than the full likelihood or the Bayesian methods, without compromising the accuracy \citep{hejase2016}.

Nevertheless, network inference is still a computationally demanding procedure. It is not uncommon to conduct the analysis with a handful of taxa, which narrows the scope of biological investigation. In this study, we present \textsc{SNaQ2}, a new version of \textsc{SNaQ} with improved computational efficiency via (1) weighted random selection of quartets, (2) parallelization of the quartet likelihood calculation during composite likelihood computation, and (3) probabilistic decision-making during network search heuristics. In the following, we first briefly introduce the essence of the original \textsc{SNaQ} followed by the key improvemetns made in \textsc{SNaQ2}. Then, we present the result of our benchmarks that compares the performance of \textsc{SNaQ} and \text{SNaQ2} and we apply \textsc{SNaQ2} on empirical datasets. Our results clearly demonstrate improved efficiency in \textsc{SNaQ2}.

%\pagebreak
\section{Methods}\label{sec2}
\subsection{Original SNaQ}\label{subsec1}
%$4 \choose 2$=3
%\textsc{SNaQ} estimates phylogenetic networks from multi-locus data using composite likelihood. 
While details of \textsc{SNaQ} is available in \cite{solis-lemus2016}, we make a brief description here for readers to clearly see the improvemetns in \textsc{SNaQ2} in the following subsection. To quantify the fit of the data on a network topolgy, \textsc{SNaQ} first extracts unrooted quarnets (i.e., networks with four tips) from the full network, and computes the expected concordance factor (CF) for each quarnet. Note CF represents the proportion of genes whose true relationship is the quartet under the coalescent model \citep{baum2007}. For a taxon set $X\in\{a,b,c,d\}$ has three possible ways to cluster four taxa into two groups of two (i.e., separated by a split \citep{chifman2014}). Thus, there are three unrooted quarnets: $q_1$ that contains $a$ and $b$ on one side and $c$ and $d$ on the other, denoted by $q_1=ab|cd$, $q_2=ac|bd$, and $q_3=ad|bc$.

Given a set of estimated gene trees $G=\{G_1,G_2,\dots,G_g\}$ for $g$ loci, the number of gene trees that match with the each of the three quarnets is denoted $X=(X_{q_1},X_{q_2},X_{q_3})$. Assuming each loci is unlinked, $X$ follows a multinomial distribution of the expected CF for each quarnet with probabilities $(CF_{q_1},CF_{q_2},CF_{q_3})$. For a level-1 network with $n\ge4$ taxa, the composite likelihood of a network is computed using:
\begin{equation}
    L=\prod_{s \in S}(CF_{q_1})^{X_{q_1}}(CF_{q_2})^{X_{q_2}}(CF_{q_3})^{X_{q_3}}
\label{eqn1}
\end{equation}
\noindent where $S$ is the collection of all quarnets exctracted from the network.

\textsc{SNaQ} heuristically searches the `best' network topology using hill climbing. Five topological `moves' to traverse the networks space and jump between different dimentions used are (i) nearest-neighbor interchange (NNI), (ii) addition of a reticulation, (iii) change direction of the reticulation edge, and move (iv) the target or (v) the origin of an existing hybridization edge. In brief, the search begins with the original topology $N_0$ that has the composite likelihood of $L_0$, and a new topology $N_1$ with the composite likelihood $L_1$ is proposed by applying randomly selected one of above five moves on $N_0$. If $L_0 < L_1$, $N_0$ is discarded and $N_1$ becomes $N_0$. Otherwise, $N_1$ is discarded and another move is applied to $N_0$. This process continues until an optima is reached. Typically, \textsc{SNaQ} executes multiple independent runs (i.e., searches). 

\subsection{Improvements in SNaQ2}\label{subsec2}
\subsubsection{Parallelization of the composite likelihood calculation}\label{subsubsec1}
While \textsc{SNaQ} utilizes parallelization mechanism by allowing each independent run on different processors (or cores) using \textsc{Julia} package \textsc{Distributed}, \textsc{SNaQ2} further improves the computational efficiency by multithreading the composite likelihood calculation. In particular, extraction of quartet topologies from a network, calculation of expected CFs of the extracted quartet, and computation of quarnet likelihood are now parallelized. This setting allows to allocate all runs independently on separate high-performance computing nodes, with each node fully utilized to parallelize the composite likelihood calcuation for the run it is responsible for.

\subsubsection{Sampling subset of quartets for composite likelihood}\label{subsubsec3}
In \textsc{SNaQ}, all quartets extracted from a network were used to compute composite likelihood. While this computation is generally efficient, it may lead to the bottleneck as the number of taxa increases, since there are $n \choose 4$ quartets in a network. In \textsc{SNaQ2}, a new argument \texttt{propQuartets}, which specifies the proportion of sampled quarnets from the full set of quarnets extracted for the composite likelihood calculation, is available in the main function \texttt{snaq!}. The value of \texttt{propQuartets} must be non-negative float $\le 1$. \textsc{SNaQ2} currently subseamples quartets in a randomized manner, which is a common subsampling approach (e.g., SVDQuartets \citep{chifman2014,chifman2015}) in phylogenetic inference, although some studies show weighted subsampling can improve accuracy (e.g., ASTRAL \citep{zhang2022b}).

\subsubsection{Topological moves using quartet weighting}\label{subsubsec2}
All topological moves, except `change direction of the reticulation edge', involve random selection of a tree edge in $N_0$ that acts as the point of moficiation. For example, to perform the topological move `move the origin of an existing hybridization edge', a reticulation edge whose head is at a randomly selected reticulation node $u$ is selected at the probability of $\gamma$ (i.e., the inhertiance probability assigned to each reticulation edge), followed by a $random$ selection of a tree edge that will have a new node $u'$ in the middle. Then, the head of the selected reticulation edge becomes $u'$ with $\gamma$ and the original incoming branch becomes the other reticulation edge with ($1-\gamma$). All nodes with degree two are removed.

This stochasticity produced by a random selection of a tree branch to make a move can result in increased computational time to find a optima during the search process. We make an improvement in heuristics by selecting the edge via weighted random sampling where the weight is $\Delta CF=\sum^3_{i=1}\left|X_{q_i}-CF_{q_i}\right|$, calculated for every quartet extracted from a network. The rationale is that if an edge (or split) occurs in the frequency that deviates from the expectation, that edge is unlikely to present in the true network topology. A new argument \texttt{probQR} is added in the function \texttt{snaq!}, which must be non-negative float $\le 1$ that defines the probability of an edge chosen based on the weight (i.e, 0.0=full random and 1.0=full weighted).


\subsection{Evaluation using simulated and empirical data}\label{subsec3}
\subsubsection{Simulation}%when $n\ge20$
We evaluate the performance of \textsc{SNaQ2} using simulation. A set of species networks that has $n=\{10, 20, 30\}$ tips with $h$ reticulations where $h=\{1,3\}$ when $n=10$, and $h=\{1,3,5\}$ otherwise are manually generated. More specifically, we generated a species tree under a Yule process using \textsc{R} package \textsc{phytools} \citep{revell2012} for each $n$, then we sequentially added reticulation onto the topology at arbitrary position. We checked each network is level-1, both manually and using \textsc{R} package \textsc{SiPhyNetworks} \citep{justison2023}, considering that both versions of \textsc{SNaQ} assume the true networks belongs to the class of level-1 networks. For each species network, we set every branch length 0.5, 1.0, or 2.0 colescent unit to represent high, medium, and low amount of incomplete lineage sorting.

Using \textsc{Julia} package \textsc{PhyloCoalSimulations} \citep{fogg2023}, we generate a set of $g\in\{300, 1000, 3000\}$ gene trees for each species network. For each gene tree, we generate multiple sequence alignment that is $10^3$ bp long using \textsc{ms} \citep{hudson2002}, setting the branch length scale parameter to 0.03 and base frequency of nucleotides as A=0.3, C=0.2, G=0.2, and T=0.3 under the HKY model. The sequence alignment is then used as an input file to estimate gene tree using \textsc{IQ-TREE} 1.6.12 \citep{nguyen2015} with the best substituion model being identified withtin the sotftware with default parameters. Gene tree estimation error is measured using \textsc{Python} package \textsc{FastMulRFS} \citep{molloy2020}.

A table of CFs computed from the set of estimated gene trees are used for network estimation using \textsc{SNaQ} and \textsc{SNaQ2}. The starting topology is randomly selected among the estimated gene trees and the true $h$ is specified. All parameters were set default and identically in both versions of \textsc{SNaQ}, but we additionally specified \texttt{probQuartets} $\in\{1.0, 0.9, 0.7\}$ and \texttt{proqQR} $\in\{0,0.5,1.0\}$ for \textsc{SNaQ2}. We recorded runtime for each network analysis. We computed the hardwired cluster dissimilarity metric between the estimated network with the true network as well as the major trees of the estimated network and the true network using \textsc{Julia} package \textsc{PhyloNetworks}. All computations are conducted using Condor at University of Wisconsin-Madison. The analyses are executed using various number of processors $\in\{4,8,16\}$ to compare the efficiency in different computing power. One hundred replicates are made.


%Add species networks used in supplementary materials 
%http://blog.phytools.org/2015/10/simulating-species-tree-from-genus-tree.html
%Add sophisticated simulation pipeline in supp.

\subsubsection{Empirical data}
Using \textsc{SNaQ2}, we reanalyze the table of CFs used in \cite{solis-lemus2016} that is obtained from the transcriptome data in \cite{cui2013} to reconstruct the evolutionary history of 24 swordtails and playfishes (\textit{Xiphophorus}: Poeciliidae). We set \texttt{probQuartets}=0.7, \texttt{proqQR}=1.0, and $h=2$ (as specified in the \cite{solis-lemus2016}). We used 16 processors.

2. find something else

\section{Results and discussion}\label{sec3}
\subsection{Simulation}\label{subsec4}
GTEE
\subsection{Empirical data}\label{subsec5}

\section{Conclusion}\label{sec4}

\section{Competing interests}
No competing interest is declared.

\section{Author contributions statement}
NK and SK designed and consucted analysis, prepared the manuscript. TC implemented improvements into SNaQ. 

\section{Acknowledgments}
WID Server? 

%\bibliographystyle{plain}
%\bibliography{reference}
%USE THE BELOW OPTIONS IN CASE YOU NEED AUTHOR YEAR FORMAT.
\bibliographystyle{abbrvnat}
\bibliography{reference}

\end{document}
