%%
%% Copyright 2022 OXFORD UNIVERSITY PRESS
%%
%% This file is part of the 'oup-authoring-template Bundle'.
%% ---------------------------------------------
%%
%% It may be distributed under the conditions of the LaTeX Project Public
%% License, either version 1.2 of this license or (at your option) any
%% later version.  The latest version of this license is in
%%    http://www.latex-project.org/lppl.txt
%% and version 1.2 or later is part of all distributions of LaTeX
%% version 1999/12/01 or later.
%%
%% The list of all files belonging to the 'oup-authoring-template Bundle' is
%% given in the file `manifest.txt'.
%%
%% Template article for OXFORD UNIVERSITY PRESS's document class `oup-authoring-template'
%% with bibliographic references
%%

%%%CONTEMPORARY%%%
\documentclass[unnumsec,webpdf,contemporary,large]{oup-authoring-template}%
%\documentclass[unnumsec,webpdf,contemporary,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,contemporary,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,contemporary,small]{oup-authoring-template}

%%%MODERN%%%
%\documentclass[unnumsec,webpdf,modern,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,webpdf,modern,medium]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,modern,small]{oup-authoring-template}

%%%TRADITIONAL%%%
%\documentclass[unnumsec,webpdf,traditional,large]{oup-authoring-template}
%\documentclass[unnumsec,webpdf,traditional,large,namedate]{oup-authoring-template}% uncomment this line for author year citations and comment the above
%\documentclass[unnumsec,namedate,webpdf,traditional,medium]{oup-authoring-template}
%\documentclass[namedate,webpdf,traditional,small]{oup-authoring-template}

%\onecolumn % for one column layouts

%\usepackage{showframe}

\graphicspath{{Fig/}}

% line numbers
%\usepackage[mathlines, switch]{lineno}
%\usepackage[right]{lineno}
\usepackage{lipsum}

\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}%
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.
\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%
\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}

\begin{document}

\journaltitle{Journal Title Here}
\DOI{DOI HERE}
\copyrightyear{2022}
\pubyear{2019}
\access{Advance Access Publication Date: Day Month Year}
\appnotes{Paper}

\firstpage{1}

%\subtitle{Subject Section}

\title[Short Article Title]{Article Title}

\author[1,$\ast$]{First Author}
\author[2]{Second Author}
\author[3]{Third Author}
\author[3]{Fourth Author}
\author[4]{Fifth Author\ORCID{0000-0000-0000-0000}}

\authormark{Author Name et al.}

\address[1]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[2]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}
\address[4]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \postcode{Postcode}, \state{State}, \country{Country}}}

\corresp[$\ast$]{Corresponding author. \href{email:email-id.com}{email-id.com}}

\received{Date}{0}{Year}
\revised{Date}{0}{Year}
\accepted{Date}{0}{Year}

%\editor{Associate Editor: Name}

%\abstract{
%\textbf{Motivation:} .\\
%\textbf{Results:} .\\
%\textbf{Availability:} .\\
%\textbf{Contact:} \href{name@email.com}{name@email.com}\\
%\textbf{Supplementary information:} Supplementary data are available at \textit{Journal Name}
%online.}

\abstract{Abstracts must be able to stand alone and so cannot contain citations to
the paper's references, equations, etc. An abstract must consist of a single
paragraph and be concise. Because of online formatting, abstracts must appear
as plain as possible.}
\keywords{keyword1, Keyword2, Keyword3, Keyword4}

% \boxedtext{
% \begin{itemize}
% \item Key boxed text here.
% \item Key boxed text here.
% \item Key boxed text here.
% \end{itemize}}

\maketitle

%Characteristics of scalability and their impact on performance.
\section{Introduction}\label{sec1}
Phylogenetic networks is a acyclic directed graph that generalizes conventional bifurcating phylogenetic trees by allowing two branches to merge at a single node and createing a reticulation. Phylogenetic networks can depict complex biological scenarios that the trees cannot, such as hybrid speciation, introgression, allopolyploid speciation and so on. A hand of computational methods that estimate phylogenetic networkrs from genomic data has been proposed, however, the wide use of networks in practice is hindered by their lack of scalability. Scalability in general refers to the ability of a system to process a growing amount of work in a decreasing or stable amount of tim (Bondi, 2000). Even worse, phylogenetic network estimation belongs to the class of NP-Complete or also NP-Hard problems (non-deterministic polynomial-time). Some attempts to ameliorate this issue has been made previously but the computational burdern is too heavy for dataset size that are typically applied to the tree analyses.

One way to improve computational efficiency in the network inference is to pre-process the genomic sequence data into a set of gene trees in prior to the network inference, with which many existing methods implements (e.g., PhyloNet, SNaQ, etc.). Computational cost can be further ameliorated by using composite likelihood (or pseudolikelihood) framework, which decomposed the large network in a set of smaller problems (e.g., trees or networks with three or four tips), excecute likelihood computation on each of them, and combine them together to approximate the likelihood of the large network. This approach has been useful in several tree (e.g., MP-EST) and network (e.g., SNaQ, PhyNEST) inference methods and shown to be much faster than computing full likelihood of a topology without compromising accuracy.

Nevertheless, network inference is still a computationally demanding. It becomes prohibitive for more than 25 taxa using SNaQ, one of the fastest method available, and in practice, it is not uncommon that a network analysis only contains a handful of species (e.g., XXX, YYY, ZZZ) which narrows the scope of biological investigation. Like many existing bioinformatic tools, it is time to make use of different parallelization strategies for network estimation in order to reduce the execution time. In this study, we present SNaQ 2, an network inference method that enhances the computational speed through multithreading and making probabilistic decisions during the inference process. We first present SNaQ and the key improvemetns made in SNaQ2. Then, we present the result of our benchmarks that compares the accuracy and efficiency between SNaQ and SNaQ2; followed by appication of SNaQ2 on empirical datasets. Based on our study, we show that SNaQ2 estimates phylogenetic networks order of magnitude faster than the original SNaQ without compromising accuracy.

\section{Methods}\label{sec2}
\subsection{Original SNaQ}\label{subsec1}
Phylogenetic networks in SNaQ are estimated from multi-locus data using pseudolikelihood. In brief, pseudolikelihood of a network is computed using the likelihood formulas of the quarnets, or four-taxon subnetworks. Since these quarnets are not independent with each other hence the likelihood calculated this approach as \textit{pseudo}-likelihood.

SNaQ utilizes the concordance factor (CF) of a quarnet that refers to the proportion of genes whose true tree displays that quartet. A split divides a quartet into two, where each side of a split contains non-overlapping two taxa, and there are $4 \choose 2$ possible splits. For example, consider a taxon set $X \in \{a,b,c,d\}$, the three splits are $q_1=ab|cd$, $q_2=ac|bd$, and $q_3=ad|bc$. Talk about G, X, CFs

\begin{equation}
    L=\prod_{s \in S}(CF_{q_1})^{X_{q_1}}(CF_{q_2})^{X_{q_2}}(CF_{q_3})^{X_{q_3}}
\end{equation}

Now we have a way to quantify the fit of data onto a phylogenetic networks, the network that maximizes pseudolikelihood is searched using hill climbing algorithm. SNaQ uses one of six `moves' to traverse the network space and jump between different dimensions. The five `moves' are (1) nearest-neighbor interchange, (2) addition (or deletion) of a reticulation, (3) change direction of the reticulation edge, and move (4) the target or (5) the origin of an existing hybridization edge. One of five moves are randomly selected at each iteration.

\subsection{Improvements in SNaQ2}\label{subsec2}
\subsubsection{Parallelization of the pseudolikelihood calculation}\label{subsubsec1}
Original SNaQ make use of parallelization mechanisms by allowing each run on different processors (or cores) using Julia package \textsc{Distributed}. SNaQ2 further improves the computational speed by multithreading pseudolikelihood calculation. In particular, extraction of quartet topologies from a network, calculation of expected CFs of the extracted quartet, calculation of delta CF (i.e., $\sum|CF_{obs_i}-CF_{exp_i}|$ where $i$ corresponds to one of three possible quartet topologies are parallelized. This setting allows to allocate all runs independently on seaprate high-performance computing nodess, with each node fully utilized to parallelize the pseudolikelihood calcuation for the run it is responsible for.

\subsubsection{Sampling subset of quartets for pseudolikelihood}\label{subsubsec3}
In the orginal SNaQ, all extracted quartets were used to compute pseudolikelihood of a network. While this computation is generally efficient, it may lead to the bottleneck as the number of taxa increases since there are $n choose 4$ quartets in a network. Several studies that also utilizes quartet has shown that subsampling some quartets lead to accurate estimation of a phylogeny (e.g., SVD Quartets). In SNaQ2, we added a new argument \texttt{propQuartets} in the network inference function \texttt{snaq!}. The argument \texttt{propQuartets} specifies the proportion of randomly sampled quartets for the pseudolikelihood calculation (i.e., nonnegative float equal to or smaller than 1). 

\subsubsection{Proposals using quartet weighting}\label{subsubsec2}
During the network search, one of fives `moves' mentioned above makes a modification in the original topology $N_0$ and propose the new topology $N_1$. All moves, except the move that changes direction of the selected reticulation edge, involves random selection of a tree edge in $N_0$ that will be modified. For example, to move the origin on an existing reticulation edge, a reticulation edge whose head is at a randomly selected reticulation node $u$ is selected at prorbability of $\gamma$, followed by a $random$ selection of a tree edge that will have a new reticulation node in the middle. This stochasticity can result in increased time requires to find the global optimum during the searching process.

We make an improvement in heuristics by selecting the edge via weighted random sampling where the weight is equal to $\Delta CF=\sum^3_{i=1}\left|X_{q_i}-CF_{q_i}\right|$ calculated for every quartet extracted from a network. It can be done for a valid edges to choose from, although not implemented in the method. The rational here is... The new variable that is added to snaq! is probQR (varied from 0.0=full random to 1.0=full weighted).






\subsection{Evaluation using simulated and empirical data}\label{subsec3}
\subsubsection{Simulation}
We evaluate the performance of SNaQ2 using simulation. 



\subsubsection{Empirical data}



species networks, gene tree sim, seq sim, gen tree est, gene tree est error 

snaq1+snaq2, estimation condition: numcores, propQR, propQ

evaluation: measure time, accruacy, using condor?

Empirical data: 1. the fish data in snaq1; 2. find something nice

\section{Results and discussion}\label{sec3}
\subsection{Simulation}\label{subsec4}
\subsection{Empirical data}\label{subsec5}

\section{Conclusion}\label{sec4}

\section{Competing interests}
No competing interest is declared.

\section{Author contributions statement}
NK and SK designed and consucted analysis, prepared the manuscript. TC implemented improvements into SNaQ. 

\section{Acknowledgments}
WID Server? 

%\bibliographystyle{plain}
%\bibliography{reference}
%USE THE BELOW OPTIONS IN CASE YOU NEED AUTHOR YEAR FORMAT.
%\bibliographystyle{abbrvnat}
%\bibliography{reference}

\end{document}
